src/components/tl/ucp/allreduce/allreduce.h

```diff
@@ -36,16 +37,105 @@ ucc_status_t ucc_tl_ucp_allreduce_init(ucc_tl_ucp_task_t *task);
 #define ALLREDUCE_TASK_CHECK(_args, _team)                                     \
     CHECK_SAME_MEMTYPE((_args), (_team));
 
+#define ALLREDUCE_PACKED_KEY_MAX_LEN 1024
+
+typedef struct ucc_tl_ucp_allreduce_sw_global_work_buf_info {
+    void *packed_src_memh;
+    void *packed_dst_memh;
+} ucc_tl_ucp_allreduce_sw_global_work_buf_info_t;
+
+typedef enum ucc_tl_ucp_allreduce_sw_buf_state
+{
+    FREE,
+    RECVING,
+    REDUCING,
+    REDUCED,
+    SENDING,
+    IDLE,
+} ucc_tl_ucp_allreduce_sw_buf_state_t;
+
+typedef struct ucc_tl_ucp_allreduce_sw_buf {
+    void                               *buf;
+    ucc_tl_ucp_allreduce_sw_buf_state_t state;
+    ucs_status_ptr_t                    ucp_req;
+    size_t                              count;
+    size_t                              bytes;
+} ucc_tl_ucp_allreduce_sw_buf_t;
+
+typedef struct ucc_tl_ucp_allreduce_sw_pipeline {
```

Comment:
maybe it's better to move all those types and function declarations to separate header, allreduce.h is for generic types